ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java

```diff
@@ -95,11 +95,27 @@ public MkArrayAggregationBuffer() {
         throw new RuntimeException("Buffer type unknown");
       }
     }
+
+    private void reset() {
+      if (bufferType == BufferType.LIST) {
+        container.clear();
```

Comment:
If I understand it, the `clear()` of ArrayList should have the same problem, right?

Reply:
@dengzhhu653 The short answer is no.

`ArrayList#clear` takes `O({the number of elements, not capacity})`. Actually, I don't reproduce the same issue with `COLLECT_LIST` at least in our environment. This is a dummy code to illustrate the behavior of `ArrayList#clear`. Even if the length of `elementsContainer` is 1 million, it only updates the first `ArrayList#size`.

```
// elementsContainer is Object[] with its length enlarged based on the maximum ArrayList#size so far
// Note that this.size is ArrayList#size, not the length of elementsContainer
for (int i = 0; i < this.size; i++) {
  elementsContainer[i] = null;
}
```

In case of HashMap, the complexity depends on its capacity. That's because `clear` doesn't know which indexes have values because of hash table's nature.

```
// table is Node[] with its length enlarged based on the maximum HashMap#size so far
// Note that table.length is not HashMap#size
for (int i = 0; i < table.length; i++) {
  table[i] = null;
}
```

Reply:
get it, thanks for the explanation!